VA-IIR 1029: Add `initial_sign_in` to profile user account section

AdamKing0126 opened the pull request:
## Summary

- I work for the VA-IIR team.  We do not own the Profile component.  We are building Onboarding functionality for new users and (verified) veterans
- Add another field, `initial_sign_in` in the `profile` section of [GET /v0/user](https://department-of-veterans-affairs.github.io/va-digital-services-platform-docs/api-reference/#/operations/user/getUser)
- This field will allow us to determine if the user is new to the VA system, and allow us to present a, "Welcome to the VA" message on the front end.

## Related issue(s)

- [IIR Issue 1029](https://github.com/department-of-veterans-affairs/va-iir/issues/1029)
- [Feature Flag Cleanup](https://github.com/department-of-veterans-affairs/vets-api/issues/18714)

## Testing done

- [x] *New code is covered by unit tests*
- *Describe what the old behavior was prior to the change*: `GET /v0/user` returned an object which had, along with other info profile data.  This data did not include an initial sign in date.
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing*
    - Specs Run :)
    - Run `vets-api` as well as `vets-website` locally, observe the response to `GET /v0/user`: the `initialSignIn` attribute is present:
<img width="1489" alt="Screenshot 2024-10-07 at 9 29 59 AM" src="https://github.com/user-attachments/assets/f5c2a21c-3417-4816-963f-e8ad868a3f64">

  - *What is the testing plan for rolling out the feature?* Once it's deployed FE devs will build functionality to examine the `initialSignIn` field and determine if a user is "new"

## What areas of the site does it impact?

`GET /v0/user/`



## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [x]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [x]  I added a screenshot of the developed feature

khenson-oddball reviewed (dismissed):
LGTM

bosawt commented on app/serializers/user_serializer.rb:
```diff
@@ -12,4 +12,11 @@ class UserSerializer
              :in_progress_forms, :prefills_available, :vet360_contact_information,
              :session, :onboarding
+
+  def account
```
We shouldn't expand usage of the `user.account` object, it's a deprecated object that will eventually be removed

bosawt commented on config/features.yml:
```diff
@@ -1398,6 +1398,10 @@ features:
     actor_type: user
     description: Conditionally display the "Welcome to VA" message to new (LOA1 or LOA3) users
     enable_in_development: true
+  veteran_onboarding_include_display_created_at_in_profile:
```
There's no need to create a flipper for this feature, it's an additional field to the user serializer so no reason to have a flipper that turns it on and off

bosawt reviewed (changes requested):
The intention of this is to have a field that shows the first time this user ever authenticated, right? I would create a function on the actual user object (in `user.rb`), call it something like `def first_sign_in`, and use the `UserAccount.created_at` value:

```
def initial_sign_in
    user_account.created_at
end
```

And in the user serializer, add this as a field in the `profile` function

AdamKing0126 commented:
I appreciate your input @bosawt - implemented your changes.  What do you think about me calling the `iso8061` function in the `initial_sign_in` function?  It does turn it into a string and therefore make it less easy to manipulate.

I'm sortof wondering if it would be better to just directly reference the `user.user_account.created_at` elsewhere in the code, as opposed to adding the new field on the user.  This was sortof my thinking earlier, trying to keep user.rb clean.  I don't know if this `initial_sign_in` adds much value...

bosawt commented on app/services/users/profile.rb:
```diff
@@ -58,7 +58,7 @@ def fetch_and_serialize_profile
 
     def account
-      { account_uuid: user.account_uuid }
+      { account_uuid: user.account_uuid, created_at: user.initial_sign_in }
```
Could you actually move this attribute into the `profile` section? Just a line like:
`initial_sign_in: user.initial_sign_in`

bosawt commented on app/models/user.rb:
```diff
@@ -48,6 +48,10 @@ def account_id
   end
 
+  def initial_sign_in
+    user_account ? user_account.created_at.iso8601 : ""
```
It's fine if you want to use the `iso8601` to format the string, but I think it makes more sense for that specifically in the serializer itself. So I would leave that part off of this function, but put it in the `users/profile.rb` file when you are defining the serialized field

AdamKing0126 replied:
Cool it looks like I don't need to apply the `iso8601` function, maybe because I did not rename the field?  Not sure.  In any case, it works!  Maybe it worked before and I just added unneeded complexity

AdamKing0126 commented:
Hey @bosawt I'm having a bit of trouble with [this test](https://github.com/department-of-veterans-affairs/vets-api/blob/56f38404198c9d2dda0575d6c44ed1620d80a62c/spec/requests/v0/user_spec.rb#L133-L145) because it's indicating that it's possible to be missing an mhv account.  If the [user is an mhv user](https://github.com/department-of-veterans-affairs/vets-api/blob/57a377513ea0962248b710e56c9b60987fada573/app/models/user.rb#L477), we retrieve the `UserVerification` which matches the `mhv_correlation_id`.  Otherwise it's none.  So it looks as if the user_account can indeed be none.

AdamKing0126 commented:
> Hey @bosawt I'm having a bit of trouble with [this test](https://github.com/department-of-veterans-affairs/vets-api/blob/56f38404198c9d2dda0575d6c44ed1620d80a62c/spec/requests/v0/user_spec.rb#L133-L145) because it's indicating that it's possible to be missing an mhv account. If the [user is an mhv user](https://github.com/department-of-veterans-affairs/vets-api/blob/57a377513ea0962248b710e56c9b60987fada573/app/models/user.rb#L477), we retrieve the `UserVerification` which matches the `mhv_correlation_id`. Otherwise it's none. So it looks as if the user_account can indeed be none.

I should say that I came across this issue because I initially had a check like this:
```
  def initial_sign_in
    user_account ? user_account.created_at : ""
  end
```

To guard against the absence of a user account.

bosawt commented:
> Hey @bosawt I'm having a bit of trouble with [this test](https://github.com/department-of-veterans-affairs/vets-api/blob/56f38404198c9d2dda0575d6c44ed1620d80a62c/spec/requests/v0/user_spec.rb#L133-L145) because it's indicating that it's possible to be missing an mhv account. If the [user is an mhv user](https://github.com/department-of-veterans-affairs/vets-api/blob/57a377513ea0962248b710e56c9b60987fada573/app/models/user.rb#L477), we retrieve the `UserVerification` which matches the `mhv_correlation_id`. Otherwise it's none. So it looks as if the user_account can indeed be none.

That's just a result of the factory being messed up. A user can't authenticate without having a `UserVerification` and `UserAccount`, and honestly I'd prefer it if a function like yours hard crashed if the user somehow got to that point without either. That way we can debug what is going on, because the user certainly should not be continuing to access authenticated routes

AdamKing0126 commented:
Final (?) updates:

A user should always have a user account, however the test suite did not supply that, so initially in `user.rb` I made a check for the presence of the user account before attempting to retrieve it's `create_at` timestamp.  I have removed this check and updated tests.

I also moved this attribute from the `account` section returned by the serializer, and placed it in the `profile` section, as per request by @bosawt.  Screenshot updated.

LindseySaari commented:
Platform will wait for @bosawt or team member approval before moving forward with review

bosawt commented on spec/requests/v0/user_spec.rb:
```diff
@@ -11,10 +11,12 @@
     before do
       allow(SM::Client).to receive(:new).and_return(authenticated_client)
       allow_any_instance_of(MHVAccountTypeService).to receive(:mhv_account_type).and_return('Premium')
+      allow_any_instance_of(User).to receive(:user_verification).and_return(mhv_user_verification)
```
I'd like to be able to get this spec working without stubbing this. Is the issue that the `user.mhv_correlation_id` is blank for the breaking spec?

As a backup, if you do:

`let(:mhv_user_verification) { create(:mhv_user_verification, backing_idme_uuid: mhv_user.idme_uuid) }`

Then that should properly create a user_verification associated to the user.

Also, you might need to add a `!` so it automatically runs the code since you might not refer to it, so:
`let!(:mhv_user_verification) ...`

AdamKing0126 replied:
Yeah I have fought over this one for a LONG time.  This suggested change causes 12 test failures

bosawt replied:
oh I'll pull the branch and try

bosawt replied:
Gave a suggestion to fix most the specs, and then to fix the remaining spec, put this inside the `with missing MHV accounts` context:
`let!(:mhv_user_verification) { create(:mhv_user_verification, backing_idme_uuid: mhv_user.idme_uuid) }`

bosawt replied:
that'll fix all the specs. 

Makes sense cause most specs have an `mhv_correlation_id`, so the `UserVerification` is matched on that. And then that single spec is a messed up MHV user with the `mhv_correlation_id` cleared out, so that uses the backup mechanism to associate the `UserVerification` which is the `backing_idme_uuid`

AdamKing0126 replied:
Did this resolve the failures for you, Trevor?  I still have 12 failures.  I just pushed my code up

bosawt reviewed (changes requested):
There's only one minor spec issue I was able to find, otherwise looks good! When that's resolved I'll approve

bosawt commented on spec/requests/v0/user_spec.rb:
```diff
@@ -11,10 +11,12 @@
     let(:mhv_user) { build(:user, :mhv) }
     let(:v0_user_request_headers) { {} }
     let(:edipi) { '1005127153' }
+    let(:mhv_user_verification) { create(:mhv_user_verification) }
```
```suggestion
    let!(:mhv_user_verification) { create(:mhv_user_verification, mhv_uuid: mhv_user.mhv_correlation_id) }
```

bosawt commented on spec/requests/v0/user_spec.rb:
```diff
@@ -11,6 +11,7 @@
     let(:mhv_user) { build(:user, :mhv) }
     let(:v0_user_request_headers) { {} }
     let(:edipi) { '1005127153' }
+    let!(:mhv_user_verification) { create(:mhv_user_verification, backing_idme_uuid: mhv_user.mhv_correlation_id) }
```
oh sorry, I think I screwed up this suggested edit, it should be `mhv_uuid: mhv_user.mhv_correlation_id` here

bosawt replied:
not `backing_idme_uuid`

AdamKing0126 replied:
IT WORKS.  omg
